theano/tensor/basic.py

```diff
@@ -5526,17 +5527,25 @@ def connection_pattern(self, node):
 
         return [[True], [False], [True]]
 
-    def grad(self, inputs, grads):
+    def L_op(self, inputs, outputs, grads):
         start, stop, step = inputs
         gz, = grads
         # start and step affect the output values
         # but the outputs are integers so there's
         # no gradient through them
         # stop does not affect the output values,
         # just the output shape, so it is disconnected
-        return [start.zeros_like(),
-                DisconnectedType()(),
-                step.zeros_like()]
+
+        if (self.dtype in discrete_dtypes) and (step > 0):
+            return [start.zeros_like(),
+                    DisconnectedType()(),
+                    step.zeros_like()]
+        else:
+            num_steps_taken = outputs[0].shape[0]  # (stop-start)/step
```

Comment:
I realize that "number of steps" is ambiguous in that case. In my mind, if the shape was 2 for instance, it meant there was 1 step taken from the starting point.
The comment is not really accurate either: if for instance `step > 0` but `start > stop`, then `outputs[0].shape[0]` will be `1` instead.